extensions/azurecore/src/azureResource/services/terminalService.ts

```diff
@@ -0,0 +1,196 @@
+
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+import axios, { AxiosRequestConfig } from 'axios';
+import * as WS from 'ws';
+
+import { IAzureTerminalService } from '../interfaces';
+import { AzureAccount, AzureAccountSecurityToken, Tenant } from '../../account-provider/interfaces';
+
+const localize = nls.loadMessageBundle();
+export class AzureTerminalService implements IAzureTerminalService {
+	private readonly apiVersion = '?api-version=2018-10-01';
+
+	public constructor(context: vscode.ExtensionContext) {
+
+	}
+
+	public async getOrCreateCloudConsole(account: AzureAccount, tenant: Tenant, tokens: { [key: string]: AzureAccountSecurityToken }): Promise<void> {
+		const token = tokens[tenant.id].token;
+		const settings: AxiosRequestConfig = {
+			headers: {
+				'Accept': 'application/json',
+				'Content-Type': 'application/json',
+				'Authorization': `Bearer ${token}`
+			},
+			validateStatus: () => true
+		};
+
+		const metadata = account.properties.providerSettings;
+
+		const userSettingsUri = this.getConsoleUserSettingsUri(metadata.settings.armResource.endpoint);
+		const userSettingsResult = await axios.get(userSettingsUri, settings);
+
+		const preferredShell = userSettingsResult.data?.properties?.preferredShellType ?? 'bash';
+		const preferredLocation = userSettingsResult.data?.properties?.preferredLocation;
+
+		const consoleRequestUri = this.getConsoleRequestUri(metadata.settings.armResource.endpoint);
+		if (preferredLocation) {
+			settings.headers['x-ms-console-preferred-location'] = preferredLocation;
+		}
+
+		const provisionResult = await axios.put(consoleRequestUri, {}, settings);
+
+		if (provisionResult.data?.properties?.provisioningState !== 'Succeeded') {
+			throw new Error(provisionResult.data);
+		}
+		const consoleUri = provisionResult.data.properties.uri;
+
+		return this.createTerminal(consoleUri, token, account.displayInfo.displayName, preferredShell);
+	}
+
+
+	private async createTerminal(provisionedUri: string, token: string, accountDisplayName: string, preferredShell: string): Promise<void> {
+		class ShellType implements vscode.QuickPickItem {
+			constructor(public readonly label: string, public readonly value: string) {
+			}
+		}
+
+		const shells = [new ShellType('PowerShell', 'pwsh'), new ShellType('Bash', 'bash'),];
+		const idx = shells.findIndex(s => s.value === preferredShell);
+
+		const prefShell = shells.splice(idx, 1);
+		shells.unshift(prefShell[0]);
+
+		let shell = await vscode.window.showQuickPick(shells, {
+			canPickMany: false,
+			placeHolder: localize('azure.selectShellType', "Select Bash or PowerShell for Azure Cloud Shell")
+		});
+
+		if (!shell) {
+			vscode.window.showErrorMessage(localize('azure.shellTypeRequired', "You must pick a shell type"));
+			return;
+		}
+
+		const terminalName = localize('azure.cloudShell', "Azure Cloud Shell") + ` ${shell} (${accountDisplayName})`;
```

Comment:
Probably want to put (Preview) on this or something until it's gone through the proper reviews